scripts/repo-sync.sh

```diff
@@ -0,0 +1,106 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+readonly HELM_URL=https://storage.googleapis.com/kubernetes-helm
+readonly HELM_TARBALL=helm-v2.12.2-linux-amd64.tar.gz
+readonly STABLE_REPO_URL=https://charts.reactiveops.com/stable/
+readonly INCUBATOR_REPO_URL=https://charts.reactiveops.com/incubator/
+readonly S3_BUCKET_STABLE=s3://charts.reactiveops.com/stable
+readonly S3_BUCKET_INCUBATOR=gs://charts.reactiveops.com/incubator
+
+main() {
+    setup_helm_client
+    authenticate
+
+    if ! sync_repo stable "$S3_BUCKET_STABLE" "$STABLE_REPO_URL"; then
+        log_error "Not all stable charts could be packaged and synced!"
+    fi
+    if ! sync_repo incubator "$S3_BUCKET_INCUBATOR" "$INCUBATOR_REPO_URL"; then
+        log_error "Not all incubator charts could be packaged and synced!"
+    fi
+}
+
+setup_helm_client() {
+    echo "Setting up Helm client..."
+
+    curl --user-agent curl-ci-sync -sSL -o "$HELM_TARBALL" "$HELM_URL/$HELM_TARBALL"
+    tar xzfv "$HELM_TARBALL"
+
+    PATH="$(pwd)/linux-amd64/:$PATH"
+
+    helm init --client-only
+    helm repo add reactiveops-stable "$STABLE_REPO_URL"
+    helm repo add reactiveops-incubator "$INCUBATOR_REPO_URL"
+}
+
+authenticate() {
+    echo "Checking AWS Authentication"
+    aws sts get-caller-identity
+}
+
+sync_repo() {
+    local repo_dir="${1?Specify repo dir}"
+    local bucket="${2?Specify repo bucket}"
+    local repo_url="${3?Specify repo url}"
+    local sync_dir="${repo_dir}-sync"
+    local index_dir="${repo_dir}-index"
+
+    echo "Syncing repo '$repo_dir'..."
+
+    mkdir -p "$sync_dir"
+    if ! aws s3 cp "$bucket/index.yaml" "$index_dir/index.yaml"; then
+        log_error "Exiting because unable to copy index locally. Not safe to proceed."
+        exit 1
+    fi
+
+    local exit_code=0
+
+    for dir in "$repo_dir"/*; do
+        if helm dependency build "$dir"; then
+            helm package --destination "$sync_dir" "$dir"
+        else
+            log_error "Problem building dependencies. Skipping packaging of '$dir'."
+            exit_code=1
+        fi
+    done
+
+    if helm repo index --url "$repo_url" --merge "$index_dir/index.yaml" "$sync_dir"; then
+        # Move updated index.yaml to sync folder so we don't push the old one again
+        mv -f "$sync_dir/index.yaml" "$index_dir/index.yaml"
+
+        aws s3 sync "$sync_dir" "$bucket"
+
+        # Make sure index.yaml is synced last
+        aws s3 cp "$index_dir/index.yaml" "$bucket"
```

Comment:
Seems like the index.yaml will be moved as part of the `aws s3 sync` above. Maybe the index shouldn't be moved into the sync directory or an exclude should be added to the sync to keep it out of the bucket until last?

Reply:
I was wondering about that myself.  The helm repo had this in there to make sure it synced last, but I'm honestly not sure why.